pkg/client/add_events.go

```diff
@@ -50,100 +50,177 @@ func (client *DataSetClient) AddEvents(bundles []*add_events.EventBundle) error
 		return fmt.Errorf("AddEvents - reject batch: %w", errR)
 	}
 
-	grouped := client.groupBundles(bundles)
+	// first, figure out which keys are part of the batch
+	seenKeys := make(map[string]bool)
+	for _, bundle := range bundles {
+		key := bundle.Key(client.Config.GroupBy)
+		seenKeys[key] = true
+	}
 
+	// then create all subscribers
+	// add subscriber for events by key
+	// add subscriber for buffer by key
 	client.addEventsMutex.Lock()
 	defer client.addEventsMutex.Unlock()
-	for key, bundles := range grouped {
-		client.addBundle(key, bundles)
+	for key := range seenKeys {
+		_, found := client.addEventsChannels[key]
+		if !found {
+			client.newBufferForEvents(key)
+
+			client.newChannelForEvents(key)
+		}
+	}
+
+	// and as last step - publish them
+	for _, bundle := range bundles {
+		key := bundle.Key(client.Config.GroupBy)
+		client.eventsEnqueued.Add(1)
+		client.addEventsPubSub.Pub(bundle, key)
 	}
 
 	return nil
 }
 
-func (client *DataSetClient) addBundle(key string, bundles []*add_events.EventBundle) {
+func (client *DataSetClient) newChannelForEvents(key string) {
+	ch := client.addEventsPubSub.Sub(key)
+	client.addEventsChannels[key] = ch
+	go (func(session string, ch chan interface{}) {
+		client.ListenAndSendBundlesForKey(key, ch)
+	})(key, ch)
+}
+
+func (client *DataSetClient) newBufferForEvents(key string) {
+	session := fmt.Sprintf("%s-%s", client.Id, key)
+	buf := buffer.NewEmptyBuffer(session, client.Config.Tokens.WriteLog)
+	client.initBuffer(buf, client.SessionInfo)
+
+	client.buffersAllMutex.Lock()
+	client.buffer[session] = buf
+	defer client.buffersAllMutex.Unlock()
+
+	// create subscriber, so all the upcoming buffers are processed as well
+	client.addEventsSubscriber(session)
+}
+
+func (client *DataSetClient) ListenAndSendBundlesForKey(key string, ch chan interface{}) {
+	client.Logger.Info("Listening to events with key",
+		zap.String("key", key),
+	)
+
 	// this function has to be called from AddEvents - inner loop
 	// it assumes that all bundles have the same key
 	getBuffer := func(key string) *buffer.Buffer {
-		buf := client.Buffer(key, client.SessionInfo)
+		buf := client.getBuffer(key)
 		// change state to mark that bundles are being added
 		buf.SetStatus(buffer.AddingBundles)
 		return buf
 	}
 
 	publish := func(key string, buf *buffer.Buffer) *buffer.Buffer {
-		client.PublishBuffer(buf)
+		client.publishBuffer(buf)
 		return getBuffer(key)
 	}
 
-	buf := getBuffer(key)
-	for _, bundle := range bundles {
-		added, err := buf.AddBundle(bundle)
-		if err != nil {
-			if errors.Is(err, &buffer.NotAcceptingError{}) {
-				buf = getBuffer(key)
-			} else {
-				client.Logger.Error("Cannot add bundle", zap.Error(err))
-				// TODO: what to do? For now, lets skip it
-				continue
-			}
-		}
+	for processedMsgCnt := 0; ; processedMsgCnt++ {
+		if msg, ok := <-ch; ok {
+			bundle, ok := msg.(*add_events.EventBundle)
+			if ok {
```

Comment:
I wonder if this function can be simplified a bit to make it a bit easier to read / understand - aka possibly reduce some levels of nesting and perhaps refactor some things in functions (https://en.wikipedia.org/wiki/Cyclomatic_complexity).